src/pyiterator.jl

```diff
@@ -29,16 +30,75 @@ end
 
     Base.done(po::PyObject, s) = ispynull(s[1])
 else
-    function Base.iterate(po::PyObject, s=_start(po))
+    """
+        PyIterator{T}(pyobject)
+
+    Wrap `pyobject::PyObject` into an iterator, that produces items of type `T`. To be more precise `convert(T, item)` is applied in each iteration. This can be useful to avoid automatic conversion of items into corresponding julia types.
+    ```jldoctest
+    julia> using PyCall
+
+    julia> l = PyObject([PyObject(1), PyObject(2)])
+    PyObject [1, 2]
+
+    julia> piter = PyCall.PyIterator{PyAny}(l)
+    PyCall.PyIterator{PyAny,Base.HasLength()}(PyObject [1, 2])
+
+    julia> collect(piter)
+    2-element Array{Any,1}:
+     1
+     2
+
+    julia> piter = PyCall.PyIterator(l)
+    PyCall.PyIterator{PyObject,Base.HasLength()}(PyObject [1, 2])
+
+    julia> collect(piter)
+    2-element Array{PyObject,1}:
+     PyObject 1
+     PyObject 2
+    ```
+    """
+    struct PyIterator{T,S}
+        o::PyObject
+    end
+
+    function _compute_IteratorSize(o::PyObject)
+        S = try
+            length(o)
+            Base.HasLength()
+        catch err
```

Comment:
```suggestion
        catch err
            if !(err isa PyError && pyisinstance(err.val, @pyglobalobjptr :PyExc_TypeError))
                rethrow()
            end
```

would be safer.

(Note: we can't rely on the existence of `__len__` since the Python object may be a wrapper/proxy object.  i.e., it could be doing something like https://github.com/GrahamDumpleton/wrapt/blob/a4aeb4c7924ca18f4e9afa97172c95b1dad1e925/src/wrapt/wrappers.py#L381)